hydra_client/connection/json_connection.py

```diff
@@ -51,20 +53,40 @@ def call(self, func_name, *args, **kwargs):
         # Add user_id to the kwargs if not given and logged in.
         if 'user_id' not in kwargs and self.user_id is not None:
             kwargs['user_id'] = self.user_id
+        else:
+            self.login()
 
         # Convert the arguments to JSON objects
         json_obj_args = list(self.args_to_json_object(*args))
 
+        try:
+            ret = func(*json_obj_args, **kwargs)
+        except Exception as e:
+            hb.db.DBSession.rollback()
+            hb.rollback_transaction()
+            raise
+
         # Call the HB function
-        ret = func(*json_obj_args, **kwargs)
-        for o in self.args_to_json_object(ret):
+
+        try:
+            json_resp = list(self.args_to_json_object(ret))
+        except ValueError as e:
+            log.warning(e)
+            json_resp = [ret]
+
+        for o in json_resp:
             if self.autocommit is True:
-                hb.db.commit_transaction()
+                try:
+                    hb.commit_transaction()
+                except Exception as e:
+                    hb.db.DBSession.rollback()
+                    hb.rollback_transaction()
```

Comment:
The `try` block at 62 above reraises the exception after rollback whereas this one doesn't.